cache.go

```diff
@@ -0,0 +1,872 @@
+		return &HTTPRangeSpec{false, offsetBegin, -1}, nil
+	case offsetEnd > -1:
+		if offsetEnd == 0 {
+			return nil, errInvalidRange
+		}
+		return &HTTPRangeSpec{true, -offsetEnd, -1}, nil
+	default:
+		// rangeString contains first and last byte positions missing. eg. "bytes=-"
+		return nil, fmt.Errorf("'%s' does not have valid range value", rangeString)
+	}
+}
+
+// String returns stringified representation of range for a particular resource size.
+func (h *HTTPRangeSpec) String(resourceSize int64) string {
+	if h == nil {
+		return ""
+	}
+	off, length, err := h.GetOffsetLength(resourceSize)
+	if err != nil {
+		return ""
+	}
+	return fmt.Sprintf("%d-%d", off, off+length-1)
+}
+func (c cacheHeader) Range() (rs *HTTPRangeSpec) {
+	if rangeHeader, ok := c.Header["Range"]; ok {
+		if len(rangeHeader) != 0 {
+			var err error
+			rs, err = parseRequestRangeSpec(strings.Join(rangeHeader, ""))
+			if err != nil {
+				return nil
+			}
+		}
+		return rs
+	}
+	return nil
+}
+
+func getPutMetadata(h http.Header) map[string]string {
+	metadata := make(map[string]string)
+	for k, v := range h {
+		key := fmt.Sprintf("%s%s", amzMetaPrefix, k)
+		metadata[key] = strings.Join(v, "")
+	}
+	return metadata
+}
+func isFresh(cacheCC, reqCC *cacheControl, lastModified time.Time) bool {
+	if cacheCC == nil && reqCC == nil {
+		return false
+	}
+	freshCache := cacheCC.fresh(lastModified)
+	freshReq := reqCC.fresh(lastModified)
+
+	if reqCC == nil {
+		return freshCache
+	}
+	if cacheCC == nil {
+		return freshReq
+	}
+	return freshCache && freshReq
+}
+
+func cacheHandler(w http.ResponseWriter, r *http.Request, b *Backend) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		clnt := b.cacheClient
+		if clnt == nil || !clnt.isCacheable(r.Method) || !clnt.isOnline() {
+			b.proxy.ServeHTTP(w, r)
+			return
+		}
+
+		sortURLParams(r.URL)
+		key := generateKey(r.URL, r.Host)
+		var opts minio.GetObjectOptions
+		for k, v := range r.Header {
+			opts.Set(k, strings.Join(v, ""))
+		}
+		reader, oi, _, cacheErr := clnt.GetObject(clnt.bucket, key, opts)
+		cacheHdrs := getCacheResponseHeaders(oi)
+		cc := parseCacheControlHeaders(cacheHdrs.Header)
+		reqCC := parseCacheControlHeaders(r.Header)
+		serveCache := false
+		if cacheErr == nil {
+			// write response headers and response output return
+			defer reader.Close()
+			if reqCC.neverCache() {
+				// for expired content, revert to backend and clear cache.
+				b.proxy.ServeHTTP(w, r)
+				return
+			}
+			serveCache = isFresh(cc, reqCC, cacheHdrs.LastModified())
+			if !serveCache {
+				// set cache headers for ETag and LastModified verification
+				if r.Header.Get(xhttp.ETag) == "" && cacheHdrs.ETag() != "" {
+					r.Header.Set(xhttp.IfNoneMatch, cacheHdrs.ETag())
+				}
+				if r.Header.Get(xhttp.LastModified) == "" && !cacheHdrs.LastModified().IsZero() {
+					r.Header.Set(xhttp.IfModifiedSince, cacheHdrs.LastModified().UTC().Format(http.TimeFormat))
+				}
+			}
```

Comment:
should there be a `return` here?

Reply:
no, IfModifiedSince and IfNoneMatch are being set to validate stale cache entry against backend further down.